File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/firestore/utils/ExceptionConverter.java
/*
 * Copyright 2022, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.firestore.utils;

import static io.flutter.plugins.firebase.firestore.FlutterFirebaseFirestorePlugin.DEFAULT_ERROR_CODE;

import android.util.Log;
import com.google.firebase.firestore.FirebaseFirestoreException;
import io.flutter.plugins.firebase.firestore.FlutterFirebaseFirestoreException;
import io.flutter.plugins.firebase.firestore.GeneratedAndroidFirebaseFirestore;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class ExceptionConverter {

  public static Map<String, String> createDetails(Exception exception) {
    Map<String, String> details = new HashMap<>();

    if (exception == null) {
      return details;
    }

    FlutterFirebaseFirestoreException firestoreException = null;

    if (exception instanceof FirebaseFirestoreException) {
      firestoreException =
          new FlutterFirebaseFirestoreException(
              (FirebaseFirestoreException) exception, exception.getCause());
    } else if (exception.getCause() != null
        && exception.getCause() instanceof FirebaseFirestoreException) {
      firestoreException =
          new FlutterFirebaseFirestoreException(
              (FirebaseFirestoreException) exception.getCause(),
              exception.getCause().getCause() != null
                  ? exception.getCause().getCause()
                  : exception.getCause());
    }

    if (firestoreException != null) {
      details.put("code", firestoreException.getCode());
      details.put("message", firestoreException.getMessage());
    }

    if (details.containsKey("code")
        && Objects.requireNonNull(details.get("code")).equals("unknown")) {
      Log.e("FLTFirebaseFirestore", "An unknown error occurred", exception);
    }

    return details;
  }

  public static void sendErrorToFlutter(
      GeneratedAndroidFirebaseFirestore.Result result, Exception exception) {
    Map<String, String> exceptionDetails = ExceptionConverter.createDetails(exception);
    result.error(
        new GeneratedAndroidFirebaseFirestore.FlutterError(
            DEFAULT_ERROR_CODE,
            exception != null ? exception.getMessage() : null,
            exceptionDetails));
  }
}

File: packages/cloud_firestore/cloud_firestore/android/src/main/kotlin/io/flutter/plugins/firebase/firestore/utils/ExceptionConverter.kt
/*
 * Copyright 2022, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.firestore.utils

import android.util.Log
import com.google.firebase.firestore.FirebaseFirestoreException
import io.flutter.plugins.firebase.firestore.FlutterError
import io.flutter.plugins.firebase.firestore.FlutterFirebaseFirestoreException
import io.flutter.plugins.firebase.firestore.FlutterFirebaseFirestorePlugin.Companion.DEFAULT_ERROR_CODE

object ExceptionConverter {
  fun createDetails(exception: Exception?): Map<String, String> {
    val details = HashMap<String, String>()

    if (exception == null) {
      return details
    }

    var firestoreException: FlutterFirebaseFirestoreException? = null

    if (exception is FirebaseFirestoreException) {
      firestoreException = FlutterFirebaseFirestoreException(exception, exception.cause)
    } else if (exception.cause is FirebaseFirestoreException) {
      val firestoreCause = exception.cause as FirebaseFirestoreException
      firestoreException =
          FlutterFirebaseFirestoreException(firestoreCause, firestoreCause.cause ?: firestoreCause)
    }

    if (firestoreException != null) {
      val code = firestoreException.code
      val message = firestoreException.message
      if (code != null) {
        details["code"] = code
      }
      if (message != null) {
        details["message"] = message
      }
    }

    if (details["code"] == "unknown") {
      Log.e("FLTFirebaseFirestore", "An unknown error occurred", exception)
    }

    return details
  }

  fun <T> sendErrorToFlutter(callback: (Result<T>) -> Unit, exception: Exception?) {
    callback(
        Result.failure(
            FlutterError(DEFAULT_ERROR_CODE, exception?.message, createDetails(exception))))
  }
}
